test/source-sink-errors/mzcompose.py

```diff
@@ -230,7 +231,103 @@ def assert_recovery(self, c: Composition) -> None:
                   FROM mz_internal.mz_source_statuses
                   WHERE name = 'source1'
                 running <null>
+                """
+            )
+        )
+
+
+@dataclass
+class KafkaSinkDisruption:
+    name: str
+    breakage: Callable
+    expected_error: str
+    fixage: Callable | None
+
+    def run_test(self, c: Composition) -> None:
+        print(f"+++ Running Kafka sink disruption scenario {self.name}")
+        seed = random.randint(0, 256**4)
+
+        c.down(destroy_volumes=True, sanity_restart_mz=False)
+        c.up("testdrive", persistent=True)
+        c.up("redpanda", "materialized", "clusterd")
+
+        with c.override(
+            Testdrive(
+                no_reset=True,
+                seed=seed,
+                entrypoint_extra=["--initial-backoff=1s", "--backoff-factor=0"],
+            )
+        ):
+            self.populate(c)
+            self.breakage(c, seed)
+            self.assert_error(c, self.expected_error)
+
+            if self.fixage:
+                self.fixage(c, seed)
+                self.assert_recovery(c)
+
+    def populate(self, c: Composition) -> None:
+        # Create a source and a sink
+        c.testdrive(
+            schema()
+            + dedent(
+                """
+                # We specify the progress topic explicitly so we can delete it in a test later,
+                # and confirm that the sink stalls. (Deleting the output topic is not enough if
+                # we're not actively publishing new messages to the sink.)
+                > CREATE CONNECTION kafka_conn
+                  TO KAFKA (
+                    BROKER '${testdrive.kafka-addr}',
+                    PROGRESS TOPIC 'testdrive-progress-topic-${testdrive.seed}'
+                  );
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn TO CONFLUENT SCHEMA REGISTRY (
+                  URL '${testdrive.schema-registry-url}'
+                  );
+
+                $ kafka-create-topic topic=source-topic
+
+                $ kafka-ingest topic=source-topic format=avro schema=${schema}
+                {"f1": "A"}
+
+                > CREATE SOURCE source1
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-source-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE NONE
+                # WITH ( REMOTE 'clusterd:2100' ) https://github.com/MaterializeInc/materialize/issues/16582
 
+                > CREATE SINK sink1 FROM source1
+                  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE DEBEZIUM
+                # WITH ( REMOTE 'clusterd:2100' ) https://github.com/MaterializeInc/materialize/issues/16582
+
+                $ kafka-verify-data format=avro sink=materialize.public.sink1 sort-messages=true
```

Comment:
Is it possible to add a test to check the contents of the progress topic?

Reply:
This might be tricky to write a test for because we store, I believe, MZ times in the progress topic. If that is the case, I'm not sure how we'd express a wildcard match in the validation here. We can maybe accomplish it with a variable in the testdrive file, but I'm not sure where we could pull it from in such a way that it wouldn't flake.

Will look at this tomorrow but lmk if you have ideas.

Reply:
This is more annoying than I thought because I don't think we publish the progress topic schema to the schema registry service (we always know what it is), but verifying the data requires a published schema. This test feels like it would be nice, but not necessary in the context of this PR--the changes to the progress topic are well tested.

How would you feel about creating a ticket and deferring this work?

Reply:
sgtm! Thanks for looking into this nevertheless. I think we can tinker with a `kafka-verify-progress-data` or something similar specific to the progress topic.

Reply:
FWIW, the progress topic doesn't have a schema in that sense... it's stored as JSON. I haven't tried this, but I would expect `kafka-verify-data format=json` to at least have access to the data. (Though the data in that topic can be unpredictable for other reasons!)